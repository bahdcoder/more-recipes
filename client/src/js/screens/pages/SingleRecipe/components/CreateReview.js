import { Link } from 'react-router';
import React, { Component } from 'react';

export default class CreateReview extends Component {
  constructor(props) {
    super(props);

    this.state = {
      review: ''
    };

    this.createReview = this.createReview.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  async createReview() {
    try {
      await this.props.createReview({
        recipeId: this.props.params.id,
        review: this.state.review
      });
    } catch (e) {
      console.log(e.response);
    }
  }

  handleInputChange(event) {
    
    this.setState({
      review: event.target.value
    });
  }

  render() {
    let createReview;
    if (this.props.checkAuth()) {
      createReview = (
        <div>
          <h3 className="mb-3 mt-3 text-muted">Leave a review</h3>
          <textarea cols={5} 
                    rows={5} 
                    className="form-control" 
                    placeholder="Leave a review for this recipe..." 
                    value={this.state.review}
                    onChange={this.handleInputChange} />
          <button className="btn btn-primary btn-sm mt-3 float-right"
                  disabled={this.state.review.length < 10}
                  onClick={this.createReview}>Save review</button>
        </div>
      );
    } else {
      createReview = (
        <h1 className="text-center text-muted"><Link to="/auth/login" style={{ textDecoration: 'none' }}>Sign in</Link> to leave a review</h1>
      );
    }


    return (
      <div>
        {/* Begin create reviews section */}
        {createReview}
        
        {/* End create review section */}
      </div>
    );
  }
}
