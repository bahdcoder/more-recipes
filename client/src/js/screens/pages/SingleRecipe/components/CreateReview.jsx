import PropTypes from 'prop-types';
import { Link } from 'react-router';
import React, { Component } from 'react';

/**
 * Create Review component
 */
class CreateReview extends Component {
  /**
   * Create review initialization
   * @param {object} props
   */
  constructor(props) {
    super(props);

    this.state = {
      review: ''
    };

    this.createReview = this.createReview.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }
  /**
   * Create a review
   * @return {null} null
   */
  async createReview() {
    await this.props.createReview({
      recipeId: this.props.params.id,
      review: this.state.review
    });

    this.setState({
      review: ''
    });
  }

  /**
   * Handle input change
   * @param {object} event
   * @return {null} null
   */
  handleInputChange(event) {
    this.setState({
      review: event.target.value
    });
  }

  /**
   * Render crearte review component
   * @return {jsx} jsx
   */
  render() {
    let createReview;
    if (this.props.checkAuth()) {
      createReview = (
        <div>
          <h3 className="mb-3 mt-3 text-muted">Leave a review</h3>
          <textarea
            cols={5}
            rows={5}
            className="form-control"
            placeholder="Leave a review for this recipe..."
            value={this.state.review}
            onChange={this.handleInputChange}
          />
          <button
            className="btn btn-primary btn-sm mt-3 float-right"
            disabled={this.state.review.length < 10}
            onClick={this.createReview}
          >Save review
          </button>
        </div>
      );
    } else {
      createReview = (
        <h1 className="text-center text-muted"><Link to="/auth/login" style={{ textDecoration: 'none' }}>Sign in</Link> to leave a review</h1>
      );
    }


    return (
      <div>
        {/* Begin create reviews section */}
        {createReview}

        {/* End create review section */}
      </div>
    );
  }
}

CreateReview.propTypes = {
  createReview: PropTypes.func.isRequired,
  params: PropTypes.objectOf(PropTypes.any).isRequired,
  checkAuth: PropTypes.func.isRequired
};

export default CreateReview;
