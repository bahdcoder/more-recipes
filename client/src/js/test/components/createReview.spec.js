/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { Link } from 'react-router';
import { mount, shallow } from 'enzyme';
import CreateReview from '../../screens/pages/SingleRecipe/components/CreateReview';

import setupTests from '../setupTests';

setupTests();

describe('The CreateReview component', () => {
  const signedInProps = {
    params: { id: 1 },
    checkAuth: () => true,
    createReview() { }
  };

  const signedOutProps = {
    params: { id: 1 },
    createReview() {},
    checkAuth: () => false
  };
  test('Should mount without crashing', () => {
    const tree = shallow(<CreateReview {...signedOutProps} />);

    expect(tree).toMatchSnapshot();
  });
  test('Should display sign in message if user is not signed in', () => {
    const tree = mount(<CreateReview {...signedOutProps} />);
    expect(tree.contains((
      <h1 className="text-center text-muted">
        <Link to="/auth/login" style={{ textDecoration: 'none' }}>Sign in</Link> to leave a review
      </h1>
    ))).toBe(true);
  });

  test('Should display textarea and createReview button if user is signed in', () => {
    const tree = mount(<CreateReview {...signedInProps} />);
    expect(tree.find('textarea').length).toEqual(1);
    expect(tree.find('button').length).toEqual(1);
  });

  test('Should call the createReview function when Save Review button is called', () => {
    const wrapper = mount(<CreateReview {...signedInProps} />);
    const spy = jest.spyOn(wrapper.instance(), 'createReview');
    wrapper.setState({ review: 'Some review right here.' });
    wrapper.find('button').simulate('click');
    expect(spy).toHaveBeenCalled();
  });
  test('Should set review state when the textarea changes', () => {
    const wrapper = mount(<CreateReview {...signedInProps} />);
    //  const spy = jest.spyOn(wrapper.instance(), 'handleInputChange');

    const review = 'This is an awesome review';
    wrapper.find('textarea').simulate('change', {
      target: { value: review }
    });
    // TODO: FIND OUT WHY THIS IS FAILING. DOESN'T MAKE ANY SENSE !!!
    //  expect(spy).toHaveBeenCalled();
    expect(wrapper.state().review).toBe(review);
  });
});
