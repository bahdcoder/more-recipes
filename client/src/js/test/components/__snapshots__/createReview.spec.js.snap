// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CreateReview component Should mount without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateReview
    checkAuth={[Function]}
    createReview={[Function]}
    params={
      Object {
        "id": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1
        className="text-center text-muted"
      >
        <Link
          onlyActiveOnIndex={false}
          style={
            Object {
              "textDecoration": "none",
            }
          }
          to="/auth/login"
        >
          Sign in
        </Link>
         to leave a review
      </h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            onlyActiveOnIndex={false}
            style={
              Object {
                "textDecoration": "none",
              }
            }
            to="/auth/login"
          >
            Sign in
          </Link>,
          " to leave a review",
        ],
        "className": "text-center text-muted",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Sign in",
            "onlyActiveOnIndex": false,
            "style": Object {
              "textDecoration": "none",
            },
            "to": "/auth/login",
          },
          "ref": null,
          "rendered": "Sign in",
          "type": [Function],
        },
        " to leave a review",
      ],
      "type": "h1",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1
          className="text-center text-muted"
        >
          <Link
            onlyActiveOnIndex={false}
            style={
              Object {
                "textDecoration": "none",
              }
            }
            to="/auth/login"
          >
            Sign in
          </Link>
           to leave a review
        </h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              onlyActiveOnIndex={false}
              style={
                Object {
                  "textDecoration": "none",
                }
              }
              to="/auth/login"
            >
              Sign in
            </Link>,
            " to leave a review",
          ],
          "className": "text-center text-muted",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Sign in",
              "onlyActiveOnIndex": false,
              "style": Object {
                "textDecoration": "none",
              },
              "to": "/auth/login",
            },
            "ref": null,
            "rendered": "Sign in",
            "type": [Function],
          },
          " to leave a review",
        ],
        "type": "h1",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
